In [ ]:
import numpy as np
np.__version__

'1.25.2'

In [ ]:
import sklearn
from sklearn.datasets import load_iris
sklearn.__version__

'1.2.2'

Neste módulo nós vamos utilizar a biblioteca `sklearn` para carregar um dataset real.

A documentação você encontra aqui: https://scikit-learn.org/stable/modules/generated/sklearn.datasets.load_iris.html#sklearn.datasets.load_iris

In [ ]:
iris_data = load_iris()

In [ ]:
iris_data

{'data': array([[5.1, 3.5, 1.4, 0.2],
        [4.9, 3. , 1.4, 0.2],
        [4.7, 3.2, 1.3, 0.2],
        [4.6, 3.1, 1.5, 0.2],
        [5. , 3.6, 1.4, 0.2],
        [5.4, 3.9, 1.7, 0.4],
        [4.6, 3.4, 1.4, 0.3],
        [5. , 3.4, 1.5, 0.2],
        [4.4, 2.9, 1.4, 0.2],
        [4.9, 3.1, 1.5, 0.1],
        [5.4, 3.7, 1.5, 0.2],
        [4.8, 3.4, 1.6, 0.2],
        [4.8, 3. , 1.4, 0.1],
        [4.3, 3. , 1.1, 0.1],
        [5.8, 4. , 1.2, 0.2],
        [5.7, 4.4, 1.5, 0.4],
        [5.4, 3.9, 1.3, 0.4],
        [5.1, 3.5, 1.4, 0.3],
        [5.7, 3.8, 1.7, 0.3],
        [5.1, 3.8, 1.5, 0.3],
        [5.4, 3.4, 1.7, 0.2],
        [5.1, 3.7, 1.5, 0.4],
        [4.6, 3.6, 1. , 0.2],
        [5.1, 3.3, 1.7, 0.5],
        [4.8, 3.4, 1.9, 0.2],
        [5. , 3. , 1.6, 0.2],
        [5. , 3.4, 1.6, 0.4],
        [5.2, 3.5, 1.5, 0.2],
        [5.2, 3.4, 1.4, 0.2],
        [4.7, 3.2, 1.6, 0.2],
        [4.8, 3.1, 1.6, 0.2],
        [5.4, 3.4, 1.5, 0.4],
        [5.2, 4.1, 1.5, 0.1],
  

Este dataset contém medidas de 4 estruturas das flores de 3 espécies de plantas.

Neste exercício, nós só vamos utilizar as medidas das 4 estruturas.

In [ ]:
iris_features = iris_data['data']

In [ ]:
iris_features

array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [ ]:
iris_features.shape

(150, 4)

👆 São 150 flores, e 4 medidas.

#### `np.percentile`

Calcule os percentis 10, 25 e 50 das 4 variáveis, ou seja, "ao longo das linhas".

In [ ]:
...

array([4.8, 2.5, 1.4, 0.2])

👆 Isto significa que 10% das plantas têm a primeira medida/variável menor ou igual a 4.8, a segunda menor ou igual a 2.5, a terceira menor ou igual a 1.4 e a quarta menor ou igual a 0.2.

In [ ]:
...

array([5.1, 2.8, 1.6, 0.3])

👆 Isto significa que 25% (1/4) das plantas têm a primeira medida menor ou igual a 5.1, a segunda menor ou igual a 2.8, a terceira menor ou igual a 1.6 e a quarta menor ou igual a 0.3.

In [ ]:
...

array([5.8 , 3.  , 4.35, 1.3 ])

Você consegue interpretar este resultado?

#### `np.quantile`

Repita o exercício anterior mas agora utilizando a função `np.quantile`. Utilize os mesmos valores para os quantis, mas lembre-se que os quantis são os percentis divididos por 100. Observe que os resultados são os mesmos, como esperado.

In [ ]:
...

array([4.8, 2.5, 1.4, 0.2])

In [ ]:
...

array([5.1, 2.8, 1.6, 0.3])

In [ ]:
...

array([5.8 , 3.  , 4.35, 1.3 ])

#### `np.median`, `np.mean`

Calcule a mediana e a média das 4 variáveis.

In [ ]:
...

array([5.8 , 3.  , 4.35, 1.3 ])

👆 A mediana é igual ao percentil 50 e ao quantil 0.5.

In [ ]:
...

array([5.84333333, 3.05733333, 3.758     , 1.19933333])

#### `np.std`, `np.var`

Calcule o desvio-padrão e a variância das 4 variáveis.

In [ ]:
iris_std = ...

In [ ]:
iris_std

array([0.82530129, 0.43441097, 1.75940407, 0.75969263])

In [ ]:
iris_var = ...

In [ ]:
iris_var

array([0.68112222, 0.18871289, 3.09550267, 0.57713289])

Utilize `np.allclose` para confirmar que a variância é igual ao desvio padrão elevado ao quadrado (`np.power`).

In [ ]:
...

True

#### `np.cov`

Determine a covariância entre as 4 variáveis. Dica: você vai precisar usar o array na sua forma transposta.

In [ ]:
...

array([[ 0.68569351, -0.042434  ,  1.27431544,  0.51627069],
       [-0.042434  ,  0.18997942, -0.32965638, -0.12163937],
       [ 1.27431544, -0.32965638,  3.11627785,  1.2956094 ],
       [ 0.51627069, -0.12163937,  1.2956094 ,  0.58100626]])

#### `np.corrcoef`

Determine o coeficiente de correlação de Pearson entre as 4 variáveis. Novamente, você vai precisar usar a forma transposta do array.

In [ ]:
...

array([[ 1.        , -0.11756978,  0.87175378,  0.81794113],
       [-0.11756978,  1.        , -0.4284401 , -0.36612593],
       [ 0.87175378, -0.4284401 ,  1.        ,  0.96286543],
       [ 0.81794113, -0.36612593,  0.96286543,  1.        ]])

#### `np.histogram`

Determine a contagem de valores e os intervalos para construir um histograma da variável de índice 2, com 5 intervalos no total.

In [ ]:
hist, bin_edges = ...

In [ ]:
hist, bin_edges

(array([50,  3, 34, 47, 16]), array([1.  , 2.18, 3.36, 4.54, 5.72, 6.9 ]))

👆 Observe que a maioria das plantas (50 no total) está na faixa mais baixa dos intervalos, com valores entre 1 e 2.18.

#### `np.bincount`

Utilize `np.bincount` para determinar a frequência de cada flor em intervalos unitários, em relação à variável de índice 2. Você vai precisar informar os valores como inteiros (método `astype`).

In [ ]:
...

array([ 0, 50,  0, 11, 43, 35, 11])

👆 Não tem nenhuma flor no intervalo 0 a 1, mas 50 flores no intervalo 1 a 2.